extensions/ql-vscode/src/debugger/debug-session.ts

```diff
@@ -0,0 +1,587 @@
+        undefined,
+        undefined,
+      );
+
+      this.state = "running";
+
+      // Send the `EvaluationStarted` event first, to let the client known where the outputs are
+      // going to show up.
+      this.sendEvent(
+        new EvaluationStartedEvent(
+          this.queryRun.id,
+          this.queryRun.outputDir.querySaveDir,
+          quickEvalContext,
+        ),
+      );
+
+      try {
+        // Report progress via the debugger protocol.
+        const progressStart = new ProgressStartEvent(
+          this.queryRun.id,
+          "Running query",
+          undefined,
+          0,
+        );
+        progressStart.body.cancellable = true;
+        this.sendEvent(progressStart);
+        let result: CoreCompletedQuery;
+        try {
+          result = await this.queryRun.evaluate(
+            (p) => {
+              const progressUpdate = new ProgressUpdateEvent(
+                this.queryRun!.id,
+                p.message,
+                (p.step * 100) / p.maxStep,
+              );
+              this.sendEvent(progressUpdate);
+            },
+            this.tokenSource!.token,
+            this.createLogger(),
+          );
+        } finally {
+          // Report the end of the progress
+          this.sendEvent(new ProgressEndEvent(this.queryRun!.id));
+        }
+        this.completeEvaluation(result);
+      } catch (e) {
+        const message = e instanceof Error ? e.message : "Unknown error";
+        this.completeEvaluation({
+          resultType: QueryResultType.OTHER_ERROR,
+          message,
+          evaluationTime: 0,
+        });
+      }
+    } finally {
+      this.disposeTokenSource();
+    }
+  }
+
+  /**
+   * Mark the evaluation as completed, and notify the client of the result.
+   */
+  private completeEvaluation(
+    result: CodeQLProtocol.EvaluationCompletedEvent["body"],
+  ): void {
+    this.lastResult = result;
+
+    // Report the evaluation result
+    this.sendEvent(new EvaluationCompletedEvent(result));
+    if (result.resultType !== QueryResultType.SUCCESS) {
+      // Report the result message as "important" output
+      const message = result.message ?? "Unknown error";
+      const outputEvent = new OutputEvent(message, "console");
+      this.sendEvent(outputEvent);
+    }
+
+    this.reportStopped();
+
+    this.queryRun = undefined;
+  }
+
+  private reportStopped(): void {
+    if (this.terminateOnComplete) {
+      this.terminateAndExit();
+    } else {
+      // Report the session as "stopped", but keep the session open.
+      this.sendEvent(new StoppedEvent("entry", QUERY_THREAD_ID));
+
+      this.state = "stopped";
+    }
+  }
+
+  private terminateAndExit(): void {
+    // Report the debugging session as terminated.
+    this.sendEvent(new TerminatedEvent());
+
+    // Report the debuggee as exited.
+    this.sendEvent(new ExitedEvent(this.lastResult!.resultType));
+
+    this.state = "terminated";
+  }
+
+  private disposeTokenSource(): void {
```

Comment:
It would be nice if we could use something similar to `DisposableObject` here so you can call `track` at the start of the `evaluate` method and `disposeAndStopTracking` at the end of that method without needing to keep track of the `tokenSource` manually.

We can't use `DisposableObject` directly since we can only extend 1 class, but keeping a list of tracked objects might be useful to copy.

Reply:
I've added a new `RunningQuery` class just to encapsulate all of this stuff in a single `Disposable`. The result seems quite a bit cleaner than before.